### ISSUE TITLE
Federation

### ISSUE THREAD

**Issue by docapotamus** [pjuu/pjuu#45]
I am opening this issue so I can keep a log of research and plans for the future implementation of this feature.

Similar to other social networking software which have tried with some success I would like to implement federation in to Pjuu.

Federation will allow different people to run the Pjuu software of their own servers and connect these together. This would allow a user at, for example pjuu.com to communicate with a user at example.com. This should allow these users to know that their data is kept at their Pjuu instance.

I have some ideas of how this should work but I will need to learn a lot more before any work can start. I would like the design of this communication to be implemented outside of the Pjuu software. This will allow alternative implementations to communicate across this network.

There is some security and privacy questions I will need to answer.

I will try and put everything I learn in to this issue as a central place before any work can start.

To implement this it would probably be best to move all of Pjuu's core functionality in to an API that the frontends would be able to use. I also need to plan this out. It will be relatively simple to achieve. I have thought about this extensively in the last year.

**Comment by docapotamus** [comment]
After much thought about federation I think it is not a good idea if the main plan for Pjuu is privacy.

I think privacy on a social network all comes down to how much you can trust the administrators of such a system. I don't think that passing your messages from a site you may trust to another site owned/administered by someone you may not know.

If you do not trust the administrators of the network you are working then you can by all means take GPL licensed software and deploy it yourself creating your own network.

I can not see the value in passing your messages about and do not think it solves the privacy issue the internet is facing.

I am going to leave this issue open as one day any comments may change my mind.

I have spoken to @mcrmonkey about this matter and believe we are both in agreement.

Your thoughts and comments are more than welcome on this subject. I do understand I could be wrong about the matter.

**Comment by docapotamus** [comment]
I'm going to re-open this as I think it may need more discussion.

**Comment by Taknamay** [comment]
I actually like federation, but I can see the problem with making it compatible with privacy. The only solution I can think of (which isn't really possible in a web browser) is to keep data secret from the server, and to only encrypt and decrypt the information in a client. This is how XMPP can be both federated and private, for example.

**Comment by docapotamus** [comment]
@Taknamay I agree. They are good. I'm going to try and re-visit this in the new year.

I think there needs to be a fine balance between owning your own data and privacy but I think it can be done :)

**Comment by pizzamaker** [comment]
+1 for federation. It's the future!

**Comment by docapotamus** [comment]
It may be worth following another federated systems protocol and trying to fit Pjuu's data model to that.

I will do more investigation.

**Comment by aetherfaerber** [comment]
As you are using Python, [pypump](http://pypump.readthedocs.io/en/latest/), which was created for [MediaGoblin](http://mediagoblin.org) seems like the logical place to start. It reimplements the protocol used by pump.io, the successor of status.net. 
Still, pump's main object seems to be data ownership, privacy or access rights is not part of the plan, iirc.

**Comment by pizzamaker** [comment]
I also suggest taking a look at the Activity Streams format, which aims to be a federation standard within the W3C: http://activitystrea.ms/

Services like MediaGoblin and GNU social are expected to move towards it. Maybe Diaspora as well.

**Comment by docapotamus** [comment]
This needs more thought since the introduction of post permissions. I doubt we are ever going to be able to trust federation for trusted posts.

If you do have a trusted follower on a remote node you can not guarantee the remote node will abide by the rules. They could post it to all followers on that node if they so wish.

This would be far to big a problem for privacy at the moment.

@aetherfaerber @pizzamaker thanks for the links though. I think we have a few blockers/design decisions to get round to before we can look at the implementation. I don't think any current federation implementation will fit. Unless we apply it to a subset of Pjuu posts.

**Comment by pizzamaker** [comment]
Yes, perhaps trusted posts can simply not follow standards, while the rest of the code base complies with it. Just an idea :)

**Comment by docapotamus** [comment]
@pizzamaker I think that would make the most sense.

**Comment by docapotamus** [comment]
After much thought and discussion on this issue I do not think federation is something we are going to offer with Pjuu.

There are plenty of federated social networks out there for people to use if they so wish.

My main concern if privacy. Shipping users data around a network doesn't make any sense. It also takes control away from the original poster as to who sees their posts, their ability to reliably delete them, and could expose their content to misbehaving network nodes.

The only thing I believe we should offer is an easier customization and deployment process. This will allow users who wish to own there own data to run their own site using the Pjuu software. This can be done now but with a lot of effort if you do not want it to have the Pjuu branding and pages.

We can also address the ownership issue in the terms and conditions, and privacy policy on https://pjuu.com. But other sites running the software would have to address these issues them selves.
